Improve Ruby string interning

nirvdrum opened the pull request:
This PR improves the Ruby string interning process by allowing look-ups using substrings without having to extract their bytes. It also switches to using TruffleString's `InternalByteArray`, allowing string interning to work on native strings (I haven't run into a case requiring this yet).

nirvdrum commented on src/main/java/org/truffleruby/core/string/TBytesKey.java:
```diff
@@ -62,4 +91,51 @@ public String toString() {
         return TruffleString.fromByteArrayUncached(bytes, encoding, false).toString();
     }
 
+    private static int hashCode(InternalByteArray byteArray) {
```
I think it would be nice if we could upstream this into Truffle's `ArrayUtils`.

nirvdrum replied:
That'd be fine, too. There had been some concerns previously about exposing public APIs in something named "internal", but I don't really care where it goes.

eregon commented on src/main/java/org/truffleruby/core/string/TStringCache.java:
```diff
@@ -69,20 +70,37 @@ private void register(TruffleString tstring, RubyEncoding encoding) {
+    }
 
+    @TruffleBoundary
+    private TruffleString getTString(TBytesKey key) {
```
Could you name this one `lookupKey` to make it clear it should only be used for lookup, not for adding in the map?

eregon commented on src/main/java/org/truffleruby/core/string/TBytesKey.java:
```diff
@@ -12,19 +12,48 @@
     private final byte[] bytes;
+    private final int offset;
+    private final int length;
+    private final boolean isImmutable;
```
I think this should be tracked externally and passed to `makeCacheable()`, to save footprint.

eregon reviewed (commented):
Right it makes sense to be able to lookup without copying bytes.
OTOH when adding an entry in the cache it should copy if not a perfect fit, to avoid holding onto extra hidden bytes "outside the substring". The PR already handles that, nice.

eregon commented:
I'm curious, did you notice the current approach was a bottleneck in some benchmark?

nirvdrum commented:
Improving `String#-@` came up while continuing working on #2089. YAML loading of i18n data adds a lot of time to our test setup. I used the script from https://github.com/oracle/truffleruby/issues/2089#issuecomment-1396360889 and profiled with that.

The savings wasn't as pronounced as I had hoped for. Looking at the SVG from `jt profile` (collected before addressing PR feedback) we have:

| Case | Self Samples % |
|--------|--------|
| Before | 4.53% |
| After | 3.95% |

In MRI, the string interning process accounts for a trivial amount of total execution. We have more going on with the concurrent collection, but I wanted to try to reduce our overhead as much as possible. Reducing memory copies looked like a straightforward improvement.

Setting aside the i18n YAML loading, I noticed that RubyGems will intern a few small substrings repeatedly. The impact of that is much harder to measure. I'm assuming doing less work will yield better results in this case.

nirvdrum commented on src/main/java/org/truffleruby/core/string/TBytesKey.java:
```diff
@@ -62,4 +91,51 @@ public String toString() {
+    }
+
+    private boolean equalBytes(TBytesKey a, TBytesKey b) {
+        if (a.isPerfectFit() && b.isPerfectFit()) {
+            return Arrays.equals(a.bytes, b.bytes);
+        }
```
Will Graal optimize both cases? `Array.equals(byte[], byte[])` is annotated as `@IntrinsicCandidate`, while the variant with specified offsets and end points is not.

eregon replied:
They both end up in `vectorizedMismatch` and that's intrinsified by Graal, yes

eregon replied:
`Arrays.equals(byte[], byte[])` is also intrinsified by Graal so maybe it's a tiny bit better.
I guess the only way to know for sure is to benchmark it.

eregon replied:
It's fine as it is, no need to spend too much time on it, either way is fine.

nirvdrum commented on src/main/java/org/truffleruby/core/string/TStringCache.java:
```diff
@@ -69,20 +70,37 @@ private void register(TruffleString tstring, RubyEncoding encoding) {
     }
 
-    @TruffleBoundary
     public TruffleString getTString(byte[] bytes, RubyEncoding rubyEncoding) {
```
I really wish we had a `reason` field on `@TruffleBoundary` because it's easy to look at this and say "there's no reason this can't be compiled, so lets do away with the boundary". I've been operating under the assumption that anything that could run without a boundary should and let Truffle's heuristics sort out the rest.

eregon replied:
Feel free to add a code comment about it. Here it's a case of "no value to PE this code, and worse due to forcing an extra allocation that might not be needed otherwise". We should only PE what can benefit from PE, there is a warmup cost to PE too much code.

eregon commented on src/main/java/org/truffleruby/core/string/TBytesKey.java:
```diff
@@ -12,19 +12,44 @@
 import org.truffleruby.core.encoding.RubyEncoding;
+import org.truffleruby.core.encoding.TStringUtils;
 
 public final class TBytesKey {
```
Could you document here or somewhere else why we support offset and length? (to avoid extra copying for substrings during lookup, but not when inserting a new entry, ref #makeCacheable)

eregon commented on src/main/java/org/truffleruby/core/string/TBytesKey.java:
```diff
@@ -62,4 +87,51 @@ public String toString() {
+        return offset == 0 && length == bytes.length;
+    }
+
+    public TBytesKey makeCacheable(boolean isImmutable) {
```
Could you document this? (it creates a perfect fit key, so it does not hold on any extra bytes that it does not need when inserting a new entry to avoid leaking bytes outside the substring)

nirvdrum commented:
Feedback address in #3216. I messed up the PR management and GitHub won't let me re-open this PR.
